common/task/sequentialTaskProcessor.go

```diff
@@ -0,0 +1,197 @@
+		logger:              logger,
+	}
+}
+
+func (t *sequentialTaskProcessorImpl) Start() {
+	if !atomic.CompareAndSwapInt32(&t.status, common.DaemonStatusInitialized, common.DaemonStatusStarted) {
+		return
+	}
+
+	t.waitGroup.Add(t.coroutineSize)
+	for i := 0; i < t.coroutineSize; i++ {
+		coroutineTaskQueue := t.coroutineTaskQueues[i]
+		go t.pollAndProcessTaskQueue(coroutineTaskQueue)
+	}
+	t.logger.Info("Task processor started.")
+}
+
+func (t *sequentialTaskProcessorImpl) Stop() {
+	if !atomic.CompareAndSwapInt32(&t.status, common.DaemonStatusStarted, common.DaemonStatusStopped) {
+		return
+	}
+
+	close(t.shutdownChan)
+	if success := common.AwaitWaitGroup(&t.waitGroup, time.Minute); !success {
+		t.logger.Warn("Task processor timeout trying to stop.")
+	}
+	t.logger.Info("Task processor stopped.")
+}
+
+func (t *sequentialTaskProcessorImpl) Submit(task SequentialTask) error {
+	hashCode := int(task.HashCode()) % t.coroutineSize
+	taskQueue := t.coroutineTaskQueues[hashCode]
+	// need to dispatch this task set
+	select {
+	case <-t.shutdownChan:
+	case taskQueue <- task:
+	}
+	return nil
+}
+
+func (t *sequentialTaskProcessorImpl) pollAndProcessTaskQueue(coroutineTaskQueue chan SequentialTask) {
+	defer t.waitGroup.Done()
+
+	for {
+		select {
+		case <-t.shutdownChan:
+			return
+		default:
+			t.batchPollTaskQueue(coroutineTaskQueue)
+		}
+	}
+}
+
+func (t *sequentialTaskProcessorImpl) batchPollTaskQueue(coroutineTaskQueue chan SequentialTask) {
+	bufferedSequentialTasks := make(map[interface{}][]SequentialTask)
+	indexTasks := func(task SequentialTask) {
+		sequentialTasks, ok := bufferedSequentialTasks[task.QueueID()]
+		if ok {
+			sequentialTasks = append(sequentialTasks, task)
+			bufferedSequentialTasks[task.QueueID()] = sequentialTasks
+		} else {
+			bufferedSequentialTasks[task.QueueID()] = []SequentialTask{task}
+		}
+	}
+
+	select {
+	case <-t.shutdownChan:
+		return
+	case task := <-coroutineTaskQueue:
+		indexTasks(task)
+	BufferLoop:
+		for i := 0; i < coroutineTaskProcessingBatchSize-1; i++ {
+			select {
+			case <-t.shutdownChan:
+				return
+			case task := <-coroutineTaskQueue:
+				indexTasks(task)
+			default:
+				// currently no more task
+				break BufferLoop
+			}
+		}
+	}
+
+	for _, sequentialTasks := range bufferedSequentialTasks {
+		t.batchProcessingSequentialTasks(sequentialTasks)
+	}
+}
+
+func (t *sequentialTaskProcessorImpl) batchProcessingSequentialTasks(sequentialTasks []SequentialTask) {
+	sort.Sort(SequentialTasks(sequentialTasks))
+
+	for _, task := range sequentialTasks {
+		t.processTaskOnce(task)
+	}
+}
+
+func (t *sequentialTaskProcessorImpl) processTaskOnce(task SequentialTask) {
+	var err error
+
+TaskProcessingLoop:
+	for {
+		select {
+		case <-t.shutdownChan:
+			return
+		default:
+			err = task.Execute()
+			err = task.HandleErr(err)
+			if err == nil || !task.RetryErr(err) {
+				break TaskProcessingLoop
+			}
+		}
+	}
+
+	if err != nil {
+		task.Nack()
+	} else {
```

Comment:
remove else and return - https://github.com/golang/go/wiki/CodeReviewComments#indent-error-flow